_source/_posts/2023-07-27-spring-graphql-react.md

```diff
@@ -0,0 +1,1794 @@
+  ALLOWED ORIGINS
+  ALLOWED WEB ORIGINS
+  TOKEN ENDPOINT AUTH
+  GRANTS               implicit, authorization_code, refresh_token, client_credentials
+
+ ▸    Quickstarts: https://auth0.com/docs/quickstart/webapp
+ ▸    Hint: Emulate this app's login flow by running `auth0 test login ***`
+ ▸    Hint: Consider running `auth0 quickstarts download ***`
+```
+
+Add the `okta-spring-boot-starter` dependency to the `build.gradle` file:
+
+__build.gradle__
+```groovy
+dependencies {
+    ...
+    implementation 'com.okta.spring:okta-spring-boot-starter:3.0.5'
+    ...
+}
+```
+
+Set the clientId, issuer, and audience for OAuth2 in the `application.properties` file:
+
+__application.properties__
+```properties
+okta.oauth2.issuer=https://{yourAuth0Domain}/
+okta.oauth2.client-id={clientId}
+okta.oauth2.audience=https://{yourAuth0Domain}/api/v2/
+```
+
+Add the client secret to the `.env` file:
+
+__.env__
+```shell
+export OKTA_OAUTH2_CLIENT_SECRET={clientSecret}
+```
+
+Add the following factory method to the class `SpringBootGraphQlApiConfig`, for requiring a bearer token for all requests:
+
+__SpringBootGraphQlApiConfig.java__
+```java
+    ...
+    @Bean
+    public SecurityFilterChain configure(HttpSecurity http) throws Exception {
+        http.oauth2ResourceServer(oauth2ResourceServer -> oauth2ResourceServer.jwt(withDefaults()));
+        return http.build();
+    }
+    ...
+```
+
+> __NOTE:__
+> The Okta Spring Boot starter provides the security auto-configuration out of the box, and the resource server configuration would not be necessary. For some reason, the Spring for GraphQL cors allowed origins configuration does not take effect without the customization above.
+
+Again, in the root folder, run the API server with:
+
+```shell
+source .env && ./gradlew bootRun
+```
+
+With HTTPie, send a request to the API server using a bearer access token:
+
+```shell
+ACCESS_TOKEN={auth0AccessToken}
+```
+
+```shell
+echo -E '{"query":"{\n    companyList(page: 20) {\n        id\n        SIC\n        name\n        status\n        category\n        companyNumber\n        countryOfOrigin\n    }\n}"}' | \
+  http -A bearer -a $ACCESS_TOKEN POST http://localhost:8080/graphql
+```
+
+> __NOTE:__
+> Follow [these instructions](https://auth0.com/docs/secure/tokens/access-tokens/get-management-api-access-tokens-for-testing) to get an access token for the Auth0 Management API, which can be used for testing the server API.
+
+### Add Auth0 Login to the React client
+
+When using Auth0 as the identity provider, you can configure the Universal Login Page for a quick integration without having to build the login forms. First, create a SPA application using the Auth0 CLI:
+
+```shell
+auth0 apps create \
+  --name "React client for GraphQL" \
+  --description "SPA React client for a Spring GraphQL API" \
+  --type spa \
+  --callbacks http://localhost:3000/callback \
+  --logout-urls http://localhost:3000 \
+  --reveal-secrets
```

Comment:
Reveal secrets isn't needed here because a SPA shouldn't have a secret.

Reply:
This prompts me for `allowed-origins`, so that should be passed in too. This command works w/o prompts:

```
auth0 apps create \
  --name "React client for GraphQL" \
  --description "SPA React client for a Spring GraphQL API" \
  --type spa \
  --callbacks http://localhost:3000/callback \
  --logout-urls http://localhost:3000 \
  --origins http://localhost:3000 \
  --web-origins http://localhost:3000
```

Reply:
Thanks, I updated the command.